### PR TITLE
[SPARK-10533] [SQL] handle scientific notation in sqlParser

### PR DESCRIPTION
**Pull request by adrian-wang**
https://issues.apache.org/jira/browse/SPARK-10533

val df = sqlContext.createDataFrame(Seq(("a",1.0),("b",2.0),("c",3.0)))
df.filter("_2 < 2.0e1").show

Scientific notation didn't work.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Just to clarify, do HiveSQL and other SQL dialects support this?

**Comment by adrian-wang** [comment]
@JoshRosen HiveQL supports this, so does MySQL.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/AbstractSparkSQLParser.scala [hunk 1]
```diff
@@ -102,8 +106,12 @@ class SqlLexical extends StdLexical {
   }
 
   override lazy val token: Parser[Token] =
-    ( identChar ~ (identChar | digit).* ^^
-      { case first ~ rest => processIdent((first :: rest).mkString) }
+    ( rep1(digit) ~ ('.' ~> digit.*).? ~ (exp ~> sign.? ~ rep1(digit)) ^^ {
+        case i ~ None ~ (sig ~ rest) =>
+          DecimalLit(i.mkString + "e" + sig.getOrElse("") + rest.mkString)
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/AbstractSparkSQLParser.scala]: Nit: `sig.getOrElse("")` can be simplified to `sig.mkString`.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 2]
```diff
@@ -169,6 +169,26 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
...
+      testData.filter("key > 0.9e+2"),
+      testData.collect().filter(_.getInt(0) > 90).toSeq)
+
+    checkAnswer(
+      testData.filter("key > 900e-1"),
+      testData.collect().filter(_.getInt(0) > 90).toSeq)
+
+    checkAnswer(
+      testData.filter("key > 900.0E-1"),
+      testData.collect().filter(_.getInt(0) > 90).toSeq)
```
Comment (liancheng) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Could you please add a case for literals like `9.e+1`? This should be accepted according to the defined parser rules.

**Comment by liancheng** [comment]
Thanks, merging to master.

**Pull request closed without merging**